golem-worker-executor-base/src/durable_host/golem/mod.rs

```diff
@@ -173,35 +170,94 @@ impl<Ctx: WorkerCtx> golem::api::host::Host for DurableWorkerCtx<Ctx> {
     }
 
     async fn mark_begin_operation(&mut self) -> anyhow::Result<OplogIndex> {
-        unimplemented!()
+        record_host_function_call("golem::api", "mark_begin_operation");
+        let begin_index = self.state.oplog_idx;
+
+        self.state.consume_hint_entries().await;
+        if self.state.is_live() {
+            self.state
+                .set_oplog_entry(OplogEntry::BeginAtomicRegion {
+                    timestamp: Timestamp::now_utc(),
+                })
+                .await;
+        } else {
+            let _ = get_oplog_entry!(self.state, OplogEntry::BeginAtomicRegion)?;
+
+            match self
+                .state
+                .lookup_oplog_entry_end_operation(begin_index)
+                .await
+            {
+                Some(end_index) => {
+                    debug!(
+                        "Worker {}'s atomic operation starting at {} is already committed at {}",
+                        self.worker_id, begin_index, end_index
+                    );
+                }
+                None => {
+                    debug!("Worker {}'s atomic operation starting at {} is not committed, ignoring persisted entries", self.worker_id, begin_index);
+
+                    // We need to jump to the end of the oplog
+                    self.state.oplog_idx = self.state.oplog_size;
+
+                    // But this is not enough, because if the retried transactional block succeeds,
+                    // and later we replay it, we need to skip the first attempt and only replay the second.
+                    // Se we add a Jump entry to the oplog that registers a deleted region.
```

Comment:
Could you explain this part? Not getting it properly

Reply:
The oplog is append only. So let's imagine that on first run, we fail in the middle of an atomic operation:

```
x = arbitrary oplog entry
B = atomic region start
F = failure
E = atomic region end

x x x B x x F
```

now we restart the worker, and replay the oplog. when reaching B, we jump to the end and start writing new entries (now `y`) but fail again:

```
x x x B x x F y y F
```

we restart it again, and when we replay we do the same, writing `z`s but now it does not fail within the block, but fails later:

```
x x x B x x F y y F z z z z E z z F
```

Then we restart it again and when we reach `B`, we find the corresponding `E` so we know the region was committed and we don't need to rerun it. 
This means that we need to act like only `z z z z` is between `B` and `E`, otherwise we go out of sync (because how it works is that the WASM code always runs thet same, and when it tries doing some side effect, we read the next oplog entry.

So we have to skip the entries between the first `x` and the last `F` in the region and we can use the same mechanism as we use for the "time travel jumps" here.